libbpf-rs/src/tc.rs

```diff
@@ -0,0 +1,291 @@
+    /// if unset upon attach, the kernel will assign a priority for the hook
+    pub fn priority(&mut self, priority: u32) -> &mut Self {
+        self.opts.priority = priority;
+        self
+    }
+
+    /// Query a hook to inspect the program identifier (prog_id)
+    pub fn query(&mut self) -> Result<u32> {
+        let mut opts = self.opts;
+        opts.prog_id = 0;
+        opts.prog_fd = 0;
+        opts.flags = 0;
+
+        let err = unsafe { libbpf_sys::bpf_tc_query(&self.hook as *const _, &mut opts as *mut _) };
+        if err != 0 {
+            Err(Error::System(errno::errno()))
+        } else {
+            Ok(opts.prog_id)
+        }
+    }
+
+    /// Attach the TcHook so that the program starts processing
+    ///
+    /// Once the hook is processing, changing the values will have no effect
+    /// unless the hook is attach()'d again (replace(true) being required)
+    ///
+    /// Users can create a second hook by changing the handle, the priority
+    /// or the attach_point and calling the attach() method again.
+    /// Beware doing this.  It might be better to Copy the TcHook and change
+    /// the values on the copied hook for easier detach()
+    ///
+    /// NOTE: Once a TcHook is attached, it, and the maps it uses,
+    ///       will outlive the userspace application that spawned them
+    ///       Make sure to detach if this is not desired
+    pub fn attach(&mut self) -> Result<Self> {
+        self.opts.prog_id = 0;
+        let err =
+            unsafe { libbpf_sys::bpf_tc_attach(&self.hook as *const _, &mut self.opts as *mut _) };
+        if err != 0 {
+            Err(Error::System(errno::errno()))
+        } else {
+            Ok(*self)
+        }
+    }
+
+    /// Detach a TcHook
+    pub fn detach(&mut self) -> Result<()> {
+        let mut opts = self.opts;
+        opts.prog_id = 0;
+        opts.prog_fd = 0;
+        opts.flags = 0;
+
+        let err = unsafe { libbpf_sys::bpf_tc_detach(&self.hook as *const _, &opts as *const _) };
+        if err != 0 && err != -(errno::Errno::ENOENT as i32) {
+            Err(Error::System(err as i32))
+        } else {
+            self.opts.prog_id = 0;
+            Ok(())
+        }
+    }
+
+    /// Destroy attached filters
+    ///
+    /// If called on a hook with an attach_point of EGRESS, will detach all EGRESS hooks
+    /// If called on a hook with an attach_point of INGRESS, will detach all INGRESS hooks
+    ///
+    /// If called on a hook with an attach_point of EGRESS|INGRESS, will destroy the clsact tc
+    /// qdisc and detach all hooks
+    ///
+    /// Will error with EOPNOTSUPP if attach_point is BPF_TC_CUSTOM
+    ///
+    /// It is good practice to query before destroying as the tc qdisc may be used
+    /// by multiple programs
+    pub fn destroy(&mut self) -> Result<()> {
+        let err = unsafe { libbpf_sys::bpf_tc_hook_destroy(&mut self.hook as *mut _) };
+        if (err == -(EINVAL as i32) || err == -(ENOENT as i32))
```

Comment:
I see EINVAL can also get returned with attach point == TC_EGRESS | TC_INGRESS, in the case where the parent of the hook is missing (in specific, I'm looking at the attach_point_to_config function in netlink.c). It seems like this shouldn't return Ok()

Reply:
My semantic intention was to return an Ok() if the hooks were "gone".  A hook with a missing parent in my opinion is semantically the same as destroyed as I don't think a EGRESS|INGRESS hook can exist without some sort of parent.

EG if destroy() is called twice, my opinion is that the second destroy should not return an error as it's semantically a no-op.

If you would prefer, I can enable the error for the EINVAL case, it just means the consumers of the API will need to pay a bit more attention to which error value they receive as an EINVAL shouldn't dramatically alter their program flow.

Reply:
Gotcha, thanks for your explanation. I personally like the idea of propagating whatever error libbpf returns - I think this is more future proof where if in the future libbpf returns -EINVAL for some other edge case, that will be automatically propagated by libbpf-rs. I'm also fine with leaving as is though, so the call is up to you :)

Reply:
As of most recent push, all errors are passed to the user.  With the exception of create().  IMHO erroring on a "double create" is very difficult for users to understand (see test_tc_double_create in test_tc.rs)